lib/msf/core/module_set.rb

```diff
@@ -304,33 +274,31 @@ def each_module_list(ary, opts, &block)
   # @return [Array<Array<String, Class>>] Array of arrays where the inner array is a pair of the module reference name
   #   and the module class.
   def rank_modules
-    self.sort_by { |pair| module_rank(*pair) }.reverse!
+    module_metadata.sort_by { |refname, _metadata| "#{module_rank(refname)}-#{refname}" }.reverse!
```

Comment:
Looks like this is super slow now; it seems like the sort block calls off to `module_rank(refname)` which then does a lookup on the module metadata - which is now behind a mutex lock and creation of an array:

```ruby
  # This file
  def rank_modules
    module_metadata.sort_by { |refname, _metadata| "#{module_rank(refname)}-#{refname}" }.reverse!
  end


  def module_rank(reference_name)
    module_metadata[reference_name].rank || Msf::NormalRanking
  end

 # Elsewhere

 def module_metadata(type)
    @mutex.synchronize do
      wait_for_load
      @module_metadata_cache.filter_map { |_, metadata| [metadata.ref_name, metadata] if metadata.type == type }.to_h
    end
  end
```

I believe this should be faster, but it's worth confirming it's semantically the same:

```ruby
  module_metadata.sort_by { |refname, metadata| [metadata.rank, refname] }.reverse!
```

---

Tested with:

```ruby
require 'benchmark/ips'

require 'msfenv'
framework = Msf::Simple::Framework.create

Benchmark.ips do |x|
  # Configure the number of seconds used during
  # the warmup phase (default 2) and calculation phase (default 5)
  x.config(:time => 15, :warmup => 4)

  # These parameters can also be configured this way
  x.time = 5
  x.warmup = 2

  x.report("rank_modules_string") do
    framework.modules.payloads.rank_modules_string.length
  end

  x.report("rank_modules_array") do
    framework.modules.payloads.rank_modules_array.length
  end

  x.report("rank_modules_array_without_lock") do
    framework.modules.payloads.rank_modules_array_without_lock.length
  end

  x.compare!
end
```

Results:

```
Warming up --------------------------------------
 rank_modules_string     1.000  i/100ms
  rank_modules_array     1.000  i/100ms
rank_modules_array_without_lock
                        15.000  i/100ms
Calculating -------------------------------------
 rank_modules_string      0.392  (± 0.0%) i/s -      3.000  in   7.716404s
  rank_modules_array      0.426  (± 0.0%) i/s -      3.000  in   7.055821s
rank_modules_array_without_lock
                        151.488  (± 5.9%) i/s -    765.000  in   5.068616s

Comparison:
rank_modules_array_without_lock:      151.5 i/s
  rank_modules_array:        0.4 i/s - 355.61x  slower
 rank_modules_string:        0.4 i/s - 386.61x  slower
```

Code diff that I tested with:

```diff
diff --git a/lib/msf/core/module_set.rb b/lib/msf/core/module_set.rb
index 175a32adc4..a6b297a22a 100644
--- a/lib/msf/core/module_set.rb
+++ b/lib/msf/core/module_set.rb
@@ -196,7 +196,6 @@ class Msf::ModuleSet < Hash
     module_metadata.keys
   end
 
-  protected
 
   # Enumerates the modules in the supplied array with possible limiting factors.
   #
@@ -268,13 +267,25 @@ class Msf::ModuleSet < Hash
   #   @return [String] type of modules
   attr_writer   :module_type
 
+  def rank_modules
+    rank_modules_array_without_lock
+  end
+
   # Ranks modules based on their constant rank value, if they have one.  Modules without a Rank are treated as if they
   # had {Msf::NormalRanking} for Rank.
   #
   # @return [Array<Array<String, Class>>] Array of arrays where the inner array is a pair of the module reference name
   #   and the module class.
-  def rank_modules
-    module_metadata.sort_by { |refname, _metadata| module_rank(refname) }.reverse!
+  def rank_modules_string
+    module_metadata.sort_by { |refname, _metadata| "#{module_rank(refname)}-#{refname}" }.reverse!
+  end
+
+  def rank_modules_array
+    module_metadata.sort_by { |refname, _metadata| [module_rank(refname), refname] }.reverse!
+  end
+
+  def rank_modules_array_without_lock
+    module_metadata.sort_by { |refname, metadata| [metadata.rank, refname] }.reverse!
   end
 
   # Retrieves the rank from a loaded, not-yet-loaded, or unloadable Metasploit Module.

```